File: Python/ExampleTests/production/Dollar.py
import decimal


class Dollar:
    """A monetary amount: never negative, never finer than a cent."""

    def __init__(self, value):
        text = str(value if value is not None else '').replace('$', '').strip()
        if text == '':
            self.amount = decimal.Decimal(0)
        else:
            try:
                self.amount = decimal.Decimal(text)
            except decimal.InvalidOperation as e:
                raise ValueError(f'Not a number: {value}') from e
        if self.amount < 0:
            raise ValueError(f'Dollar amount cannot be negative: {value}')
        if -self.amount.as_tuple().exponent > 2:
            raise ValueError(
                f'Dollar amount must not have more than two decimal digits: {value}')

    def plus(self, other: 'Dollar') -> 'Dollar':
        return Dollar(self.amount + other.amount)

    def minus(self, other: 'Dollar') -> 'Dollar':
        return Dollar(self.amount - other.amount)

    def times(self, factor: int) -> 'Dollar':
        return Dollar(self.amount * factor)

    def percent_of(self, percentage) -> 'Dollar':
        """The given percentage of this amount, rounded to the nearest cent."""
        raw = self.amount * percentage.to_decimal() / decimal.Decimal(100)
        return Dollar(raw.quantize(decimal.Decimal('0.01'),
                                   rounding=decimal.ROUND_HALF_UP))

    def to_decimal(self) -> decimal.Decimal:
        return self.amount

    def __eq__(self, other):
        if not isinstance(other, Dollar):
            return NotImplemented
        return self.amount == other.amount

    def __hash__(self):
        return hash(self.amount)

    def __str__(self):
        return f'Dollar{{{self.amount}}}'

File: Rust/ExampleTests/production/dollar.rs
/// A monetary amount: never negative, never finer than a cent.
/// Held as an integer number of cents so the arithmetic is exact.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default)]
pub struct Dollar {
    cents: i64,
}

impl Dollar {
    /// Reads "$10.00", "10", "0.01" and the like.
    pub fn parse(value: &str) -> Result<Self, String> {
        let text = value.replace('$', "");
        let text = text.trim();
        if text.is_empty() {
            return Ok(Dollar { cents: 0 });
        }
        let negative = text.starts_with('-');
        let digits = text.trim_start_matches(['-', '+']);
        let mut parts = digits.splitn(2, '.');
        let whole = parts.next().unwrap_or("");
        let frac = parts.next().unwrap_or("");
        if whole.is_empty() && frac.is_empty() {
            return Err(format!("Not a number: {value}"));
        }
        if !whole.chars().all(|c| c.is_ascii_digit())
            || !frac.chars().all(|c| c.is_ascii_digit())
        {
            return Err(format!("Not a number: {value}"));
        }
        if frac.len() > 2 {
            return Err(format!(
                "Dollar amount must not have more than two decimal digits: {value}"));
        }
        let whole_v: i64 = if whole.is_empty() { 0 } else {
            whole.parse().map_err(|_| "overflow".to_string())?
        };
        let frac_v: i64 = match frac.len() {
            0 => 0,
            1 => frac.parse::<i64>().map_err(|_| "bad".to_string())? * 10,
            _ => frac.parse::<i64>().map_err(|_| "bad".to_string())?,
        };
        let cents = whole_v * 100 + frac_v;
        if negative && cents != 0 {
            return Err(format!("Dollar amount cannot be negative: {value}"));
        }
        Ok(Dollar { cents })
    }

    pub fn from_cents(cents: i64) -> Self { Dollar { cents } }
    pub fn cents(&self) -> i64 { self.cents }

    pub fn plus(&self, other: &Dollar) -> Dollar { Dollar { cents: self.cents + other.cents } }
    pub fn minus(&self, other: &Dollar) -> Dollar { Dollar { cents: self.cents - other.cents } }
    pub fn times(&self, factor: i32) -> Dollar { Dollar { cents: self.cents * factor as i64 } }

    /// The given percentage of this amount, rounded half up to the nearest cent.
    pub fn percent_of(&self, percentage: &super::Percentage) -> Dollar {
        Dollar { cents: (self.cents * percentage.value() as i64 + 50) / 100 }
    }
}

impl std::fmt::Display for Dollar {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}.{:02}", self.cents / 100, (self.cents % 100).abs())
    }
}
